### PR TITLE
feat: make filesystem injectable

### PR DESCRIPTION
**Pull request by zetlen**
- User can now inject a `fs` object into an instance of `jsonfile`, to do the same operations on anything that follows the `fs` contract, (such as a Webpack virtual filesystem, for instance).
- Default export is now an instance of the new `JsonFile` constructor, initialized with the default `fs` object. It should work exactly the same as before!
- Tests added for this mode of operation.

Made this to support https://github.com/jprichardson/node-fs-extra/pull/733 to enable the same thing.

### CODE REVIEW

**Comment by zetlen** [comment]
(Couldn't get today's version of `standard` to output the exact same indentation as this originally had, so it might be easier to use GitHub's "Hide whitespace changes" to see the actual changeset.)

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,5 @@
 node_modules/
 /.project
+/.vscode
+package-lock.json
+yarn.lock
```
Comment (RyanZim) [inline .gitignore]: Remove these changes; they're not related to the issue at hand.

File: index.js [hunk 2]
```diff
@@ -1,137 +1,145 @@
...
+let realFs
 try {
-  _fs = require('graceful-fs')
+  realFs = require('graceful-fs')
 } catch (_) {
-  _fs = require('fs')
+  realFs = require('fs')
 }
 const universalify = require('universalify')
+const memoize = require('memoize-weak')
```
Comment (RyanZim) [inline index.js]: Why is this needed?

File: index.js [hunk 3]
```diff
@@ -1,137 +1,145 @@
...
+    writeFileSync
+  }
+})
 
-const jsonfile = {
-  readFile,
-  readFileSync,
-  writeFile,
-  writeFileSync
+function JsonFile (fs = realFs) {
```
Comment (RyanZim) [inline index.js]: I'm curious why a function/factory pattern was used here instead of an actual `class`?

File: .gitignore (hunk 1 above)
Comment (zetlen) [reply .gitignore]: Done.

File: index.js (hunk 2 above)
Comment (zetlen) [reply index.js]: The `jsonFileFactory` function is safe and idempotent. It doesn't do anything to the passed `fs` object, it just retains it in its closure. So it's safe to memoize for performance reasons, and the `memoize-weak` library does that without risking memory leaks.

File: index.js (hunk 3 above)
Comment (zetlen) [reply index.js]: A class would have nothing to do here, because it would have nothing on its prototype. This is equivalent, but if a class looks better for the codebase, I can refactor it.

File: index.js (hunk 2 above)
Comment (RyanZim) [reply index.js]: Have we actually run benchmarks here on the performance benefits? My guess is that there isn't enough computation being done on each call to warrant memoization.

Comment (zetlen) [reply index.js]: No benchmarks! It’s a habit of mine to memoize when I’m thinking about a performance-critical path, like the FS in a Webpack compiler, but I have no proof, and I can remove it if you want.

Comment (RyanZim) [reply index.js]: Yeah, let's remove. Given that this is used by `fs-extra`, we try to keep deps light, given the number of things that use it.

Comment (zetlen) [reply index.js]: Done.
